### PR TITLE
#3570 - Software i2c crashing ESP

### PR DESCRIPTION
**Pull request by willz1200**
If software i2c is initialised before the arduino setup function (e.g. In a class constructor) it crashes due to Exception (28). flush() is conflicting with something external to Wire.h and causing the crash. Its unnecessary because the buffers are initialised to 0 anyway.

### CODE REVIEW

**Comment by igrr** [comment]
Interesting, i thought that class member would take precedence over global (which would have to be called via `::flush`). Can you check if replacing `flush();` with `this->flush();` solves the issue? The reason why i think we shouldn't remove it is that `begin` can, in theory, be called multiple times.

**Comment by willz1200** [comment]
I tried replacing `flush()` with `this->flush()` but the esp is still crashing. Any other thoughts?

**Comment by devyte** [comment]
Access of a global object during construction of a different global object declared in a different TU is undefined behavior. The reason is that there is no guarantee in the order of construction or initialization, because the order of linking of the TUs is non deterministic.

The only case where the order is assured is in construction of globals within the same TU.

In Arduino world, global singletons is a common programming pattern. In such cases, where one object may need to init another, the init must be deferred, and can't be done on construction. To be clear, calling begin() of a global object from the constructor of a different object in a different TU may crash. In fact, calling most methods of a global object from the constructor of another object may crash.  
This is the main reason behind the begin() methods, which are supposed to be called from setup(). At that time, all globals are already constructed, so it's ok to init them in whatever order required.
In other words, your global class should also have a begin() or start() method that is called from setup(), and that method should call Wire.begin().

Closing due to user error.
